### ISSUE TITLE
Support ASP.NET Core Generic Hosts

### ISSUE THREAD

**Issue by loic-sharma** [natemcmaster/CommandLineUtils#134]
[ASP.NET Core Generic Hosts](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-2.1) are a great way to handle background tasks. Unfortunately, Generic Hosts don't do anything fancy for command line parsing. That's where CommandLineUtils comes in!

The basic idea is that CommandLineUtils should parse the command line arguments to figure out which `IHostedService` to register. 

# Attribute Pattern

In the attribute pattern, I suggest that CommandLineUtils registers the matched type using the host's `AddHostedService` if the matched type is an `IHostedService`. For example:

```csharp
public class Program : IHostedService
{
    public static int Main(string[] args)
        => await new HostBuilder()
            .UseCommandLineApplication<Program>(args) // CommandLineUtils extension!
            .RunConsoleAsync();

    [Option(Description = "The subject")]
    public string Subject { get; }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        var subject = Subject ?? "world";
        Console.WriteLine($"Hello {subject}!");
    }

    public Task StopAsync(CancellationToken cancellationToken) => Task.CompletedTask;
}
```

# Fluent apps

For fluent apps, I suggest we add some extensions on `IHostBuilder`:

```csharp
public class Program
{
    public static Task Main(string[] args)
    {
        await new HostBuilder()
            .UseCommandLineApplication(args, (hostContext, services, app) =>
            {
                app.HelpOption();
                var optionSubject = app.Option("-s|--subject <SUBJECT>", "The subject", CommandOptionType.SingleValue);

                // (1/3) "Traditional" approach:
                app.OnExecute(services =>
                {
                    var subject = optionSubject.HasValue()
                        ? optionSubject.Value()
                        : "world";

                    // This is a little icky. We pass the subject option through the service's constructor:
                    services.AddHostedService<HelloService>(provider => new HelloService(subject));
                });

                // (2/3) Alternate approach:
                // This creates a delegate-based `BackgroundService` that captures local options
                app.RunBackgroundService(CancellationToken token =>
                {
                    var subject = optionSubject.HasValue()
                        ? optionSubject.Value()
                        : "world";

                    Console.WriteLine($"Hello {subject}!");
                });

                // (3/3) Stretch goal:
                // Something else I'd like are short-lived commands that do a task
                // and then automatically shut down the .NET Generic Host after completion
                app.RunCommand(CancellationToken token =>
                {
                    var subject = optionSubject.HasValue()
                        ? optionSubject.Value()
                        : "world";

                    // After this returns, the host is shutdown gracefully.
                    Console.WriteLine($"Hello {subject}!");
                });
            })
            .RunConsoleAsync();
    }
}

public class HelloService : BackgroundService
{
    private readonly string _subject;

    public HelloService(string subject)
    {
        _subject = subject;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        Console.WriteLine($"Hello {_subject}!");
    }
}
```

**Comment by loic-sharma** [comment]
Okay, I updated my fluent app example.

One thing that I'm missing is how I could integrate a `IWebHost` with the generic `IHost` for an app that launches either a web server or a background service depending on the CLI arguments. I see that ASP.NET has a [`GenericWebHost`](https://github.com/aspnet/Hosting/blob/f9d145887773e0c650e66165e0c61886153bcc0b/samples/GenericWebHost/WebHostExtensions.cs) example - is this fleshed out enough that one could use this in production?

**Comment by natemcmaster** [comment]
I really like this idea. I don't have time to work on it right now, so I'll mark it as up for grabs. If you want to get started on an implementation, let me know.

**Comment by lucastheisen** [comment]
Ok, I think I have this [working like the OP wants](https://github.com/pastdev/dotnet-cli).  Specifically, you can:
```
using McMaster.Extensions.CommandLineUtils;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Pastdev.Cli
{
    public class Program
    {
        public static async Task<int> Main(string[] args)
        {
            return await new HostBuilder()
                .ConfigureLogging((context, builder) =>
                {
                    builder.AddConsole();
                })
                .RunCliAsync<Program>(args);
        }

        [Option(Description = "The subject")]
        public string Subject { get; }

        [Option(ShortName = "n")]
        public int Count { get; }

        private int OnExecute()
        {
            var subject = Subject ?? "world";
            for (var i = 0; i < Count; i++)
            {
                Thread.Sleep(2000);
                Console.WriteLine($"Hello {subject}!");
            }
            return 3;
        }
    }
}
```
The entire implementation is only a handful of classes...  I can generate a pull request if you are interested in this approach?  Let me know.

**Comment by natemcmaster** [comment]
Yes, I would be very interested in a pull request for this. I read your sample code and it works as I would expect it to. I may tweak a few things, like namespaces, but otherwise it looked good.

Because this adds a bunch of dependencies on more Microsoft.\* packages, I'd like this to be a separate package users can opt-in to. Maybe we call it `McMaster.Extensions.Hosting.CommandLine`? I'm bad at naming things so I'm open to other suggestions.

**Comment by lucastheisen** [comment]
@natemcmaster , i submitted the [pull req](https://github.com/natemcmaster/CommandLineUtils/pull/167).  I created it as a separate package using the namespace you suggested (i think it is the correct namespace as it is basically an extension of Hosting that adds the CLI stuff, though the inverse is also true).  I updated the dependency injection documentation to include a section for using generic host, and added a sample.  However, I did not have any unit tests, so if that is necessary, i will have to figure out how to make that happen.  Anyway, let me know what you think.
